### PR TITLE
HIDDEN_JUNCTION w/devel data; graph gen out of errorcheck

### LINKED ISSUE #433: VISIBLE_HIDDEN_COLOC & HIDDEN_JUNCTION w/only devel systems
This just randomly occurred to me, not even looking at the code...
* VISIBLE_HIDDEN_COLOC check is performed in the `HGVertex` ctor.
* HIDDEN_JUNCTION check is performed during edge compression.

All of this is after the `HighwayGraph` ctor has been fed a list of every unique point that `is_or_colocated_with_active_or_preview()`, meaning these errors won't be flagged when only devel systems are involved.

* [x] (#435) Move VISIBLE_HIDDEN_COLOC check into `WaypointQuadtree::graph_points`
* [ ] HIDDEN_JUNCTION will require some more thought.

### PR DESCRIPTION
**Pull request by yakra**
Happy St. Tib's Day!
**TLDR:**
* HIDDEN_JUNCTION datacheck takes data from devel systems into account now.
  I wanted to get this working before doing any work on railway datachecks.
* Datachecks out of graph gen = no more graph gen in errorcheck mode. **Way faster.**
  Plus fixed some errorchecks that weren't done in errorcheck mode before. Oops.

Closes #433
Closes yakra#242

Before we really get rolling, a couple minor commits:

### 1dd4799 label_references_route cleanup
cleans up an obscure canonical_waypoint_name function, making it more readable IMO.
Compiles into slightly fewer machine commands too.
Only tangentially related to the main focus of this pull request.

### 3bcedfc unexpected_designation outside graph gen
There's an unused datacheck I like to uncomment sometimes to see what I can see.
This enables us to keep doing so after 387c99b, while adding support for devel system data, and speeding up CWN a tiny bit.
This commit is used as the basis for future commits' speed comparisons, even though it doesn't affect the same tasks they do.

### CODE REVIEW

**Comment by yakra** [comment]
### a1d59f8 HIDDEN_JUNCTION w/devel data
adds support for devel system data to the HIDDEN_JUNCTION datacheck.
The idea was just to get it fully working within the existing framework, and not worry about anything for railroads yet.
* The bad news:
  * It gets [moved out](https://github.com/TravelMapping/DataProcessing/commit/a1d59f8aa439b6c61efaa2b3acb943a5da5dedb6#diff-0a73788eb5d0f37cba3d1dd9c571f3760d4f99741935edfbdea647c3d0be9f9d) of the graph edge compression process, where the necessary data was at our fingertips.
    [If a certain variable was >2](https://github.com/TravelMapping/DataProcessing/commit/a1d59f8aa439b6c61efaa2b3acb943a5da5dedb6#diff-0a73788eb5d0f37cba3d1dd9c571f3760d4f99741935edfbdea647c3d0be9f9dL86), [flag an error](https://github.com/TravelMapping/DataProcessing/commit/a1d59f8aa439b6c61efaa2b3acb943a5da5dedb6#diff-0a73788eb5d0f37cba3d1dd9c571f3760d4f99741935edfbdea647c3d0be9f9dL87-L89); done.
  Comp
  * Now, we have to redo some of our work in finding the number of incident edges.
  Intg_hj
* The good news:
  * The impact is minimized. This is only done for hidden points. Fitting into the existing framework, slap an [else](https://github.com/TravelMapping/DataProcessing/commit/a1d59f8aa439b6c61efaa2b3acb943a5da5dedb6#diff-536c689d3ed88fc7df728fd3ba1adde073950fe706db8caa6463452a01f362d5R30) onto an [if](https://github.com/TravelMapping/DataProcessing/commit/a1d59f8aa439b6c61efaa2b3acb943a5da5dedb6#diff-536c689d3ed88fc7df728fd3ba1adde073950fe706db8caa6463452a01f362d5R27), call a [function](https://github.com/TravelMapping/DataProcessing/blob/a1d59f8aa439b6c61efaa2b3acb943a5da5dedb6/siteupdate/cplusplus/classes/Waypoint/Waypoint.cpp#L348-L364), immediately [return](https://github.com/TravelMapping/DataProcessing/commit/a1d59f8aa439b6c61efaa2b3acb943a5da5dedb6?diff=unified&w=0#diff-8c5895bd8f22ab5ab0a459d74fa0c5f71115c733e35e706b78f9f1ba995a1f43R350) if `this` point isn't at the front of its colocation list.
    [The meat of the function](https://github.com/TravelMapping/DataProcessing/blob/a1d59f8aa439b6c61efaa2b3acb943a5da5dedb6/siteupdate/cplusplus/classes/Waypoint/Waypoint.cpp#L351-L360), where we iterate colocated points and build a wee data structure, only happens 38003 times (well, a bit more with the most recent repos, but you get the idea). Not bad.
  hj
  * It gets moved *into* a multi-threaded part of the program.
    Throw enough threads at the task, and the extra time approaches 0.
  hjD

**Comment by yakra** [comment]
### 7b432be VISIBLE_HIDDEN_COLOC via route_integrity
moves the VISIBLE_HIDDEN_COLOC from graph setup to `route_integrity`, where HIDDEN_JUNCTION is handled.
It's pretty much the same mechanism in both cases -- iterate through a colocation list & check whether each point's visibility differs from the first point's.
vhc

All things being equal, the time saved during graph setup should equal the additional time while `Creating label hashes and checking route integrity` with 1 thread.
But not all things are equal.
* Setup speed is down 9 ms on bsdlab, 6 ms on lab4, 2-3 ms on the other machines.
  Setup
* route_integity is up .08 s on bsdlab, 0.11 s on lab3&4, .06-.07 s on the others.
  Intg_vhc

Again, throw enough threads at it and the difference approaches 0.
vhcD

LOL, I'm sure nobody else in the real world cares about that tiny a slowdown. But I can't help it; this crap fascinates me. :grinning:
Why the different differences?
* What I can come up with is cache hits/misses.
  Before, `graph_points` iterated through the colocation list, got its data into cache, and then iterated again [two](https://github.com/yakra/DataProcessing/commit/7b432be8f15333462732e0e2b9a10430b3e0115f?diff=unified&w=0#diff-c045521fca87e318ca424d0aaa445af15b89af08da633432029f96c8ec293dd7L223) [more](https://github.com/yakra/DataProcessing/commit/7b432be8f15333462732e0e2b9a10430b3e0115f?diff=unified&w=0#diff-c045521fca87e318ca424d0aaa445af15b89af08da633432029f96c8ec293dd7L226) times. The 1st loop accessed [is_hidden](https://github.com/yakra/DataProcessing/blob/387c99b0c260938c870457bb9793d4f78b09987f/siteupdate/cplusplus/classes/Waypoint/Waypoint.h#L32) & the other 2 accessed [route](https://github.com/yakra/DataProcessing/blob/387c99b0c260938c870457bb9793d4f78b09987f/siteupdate/cplusplus/classes/Waypoint/Waypoint.h#L23) -- at the end of one Waypoint and the beginning of the next respectively, so likely to be on the same cache line the vast majority of the time.
  When the first loop was removed, most of the time "saved" was restored by the other loops taking up the slack & bringing the same data into cache from RAM.
* In contrast, all this iterating through the colocation lists during `route_itegrity` is new. ...Well, [for visible points](https://github.com/yakra/DataProcessing/commit/7b432be8f15333462732e0e2b9a10430b3e0115f?diff=unified&w=0#diff-536c689d3ed88fc7df728fd3ba1adde073950fe706db8caa6463452a01f362d5) at least. [Hidden points were in the last commit](https://github.com/yakra/DataProcessing/commit/a1d59f8aa439b6c61efaa2b3acb943a5da5dedb6#diff-8c5895bd8f22ab5ab0a459d74fa0c5f71115c733e35e706b78f9f1ba995a1f43R352), not even contributing to the time increase noted here. I'll just asuume the impact is bigger 'cause there's a lot more visible points than hidden ones. :grin:

**Comment by yakra** [comment]
### Combined effects of the last 2 commits:
closes yakra#281
Both
BothD
The upside of this minor slowdown is, we've finally gotten all datachecks out of the graph generation process.
We're all set to significantly speed up `--errorcheck` mode by disentangling it from graph generation.

**Comment by yakra** [comment]
### 387c99b graph fixes for -e mode
* Now that 7b432be's removed the last datachecks from graph generation, `--errorcheck` mode does stuff it doesn't need to -- setting up the graph structure, a heavyweight, single-threaded operation.
* Meanwhile, it doesn't do stuff it does need to -- `ErrorList` checks for invalid data in the [subgraph CSVs](https://github.com/TravelMapping/HighwayData/tree/master/graphs).

This fixes all that.
* [CSV parsing and the variables needed to get it going](https://github.com/yakra/DataProcessing/commit/387c99b0c260938c870457bb9793d4f78b09987f#diff-a372271ede6b20e80790b84ea58836ff76c7c9075f72484311c69f73bd4d2800) have been [carved out](https://github.com/yakra/DataProcessing/commit/387c99b0c260938c870457bb9793d4f78b09987f?diff=split&w=0#diff-ca5994bf1d47e51f664cfea7871bb0aff4c7ece9c5c822868b3128527d0e6cbc) of graph_generation.cpp and [moved](https://github.com/yakra/DataProcessing/commit/387c99b0c260938c870457bb9793d4f78b09987f#diff-bf9dfd5f32b1debc5f3fe76196ca17381126f15131e4b8002ffb92b15d0d57e8R186-R187) before the check to abort if there are `ErrorList` errors, where it takes all of 2 ms on my ancient desktop. :)
* This will also come in handy [in the future](https://github.com/yakra/DataProcessing/issues/251).
* This means a big cleanup for [the subgraph routines](https://github.com/yakra/DataProcessing/tree/387c99b0c260938c870457bb9793d4f78b09987f/siteupdate/cplusplus/tasks/subgraphs). Instead of siteupdateST's "prepare each category of graphs and then write their files" approach (left over from the early "translate from Python" days), it follows multi-threaded siteupdate's lead & [plows thru the whole `GraphListEntry` vector in one go](https://github.com/yakra/DataProcessing/commit/387c99b0c260938c870457bb9793d4f78b09987f#diff-ca5994bf1d47e51f664cfea7871bb0aff4c7ece9c5c822868b3128527d0e6cbcR29-R32). All that useless duplicated code is gonzo.
* Then the big time-saver: checking `if (Args::skipgraphs || Args::errorcheck)` [before](https://github.com/yakra/DataProcessing/commit/387c99b0c260938c870457bb9793d4f78b09987f#diff-bf9dfd5f32b1debc5f3fe76196ca17381126f15131e4b8002ffb92b15d0d57e8R205) constructing the graph structure, not [after](https://github.com/yakra/DataProcessing/commit/387c99b0c260938c870457bb9793d4f78b09987f#diff-ca5994bf1d47e51f664cfea7871bb0aff4c7ece9c5c822868b3128527d0e6cbcL17).
* Other minor efficiency improvements, cleanup, and behind-the-scenes work to get stuff going.

ErCh

**Comment by yakra** [comment]
**Deleted branch**
HJR5 @ `a2815fc` on BiggaTomato
interactive rebase of [`4eb43ad`](https://github.com/yakra/DataProcessing/issues/275)

**Comment by yakra** [comment]
Ready to rock. :)

**Comment by yakra** [comment]
New commit pushed: update siteupdate copyright to 2024

**Comment by yakra** [comment]
Superseded by #617.
